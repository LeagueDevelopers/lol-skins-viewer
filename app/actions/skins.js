import type { Summoner } from 'types';
import { getData } from 'api/SkinsController';

export function getSkins (port: number, password: string, summoner: Summoner) {
  return dispatch =>
    getData(port, password, summoner)
    .then(data => {
      if (data && data.champions && data.ownedChampionIds) {
        return dispatch({
          type: 'RECEIVE_SKINS',
          payload: data
        });
      }
      return dispatch({
        type: 'ERROR',
        payload: data,
        error: true
      });
    });
}

export function changeNameFilter (nextValue) {
  return {
    type: 'SKINS_FILTER_NAME_CHANGE',
    payload: nextValue
  };
}

export function changeShowFilter (nextValue) {
  return {
    type: 'SKINS_FILTER_SHOW_CHANGE',
    payload: nextValue
  };
}

export function changeSortMethod (nextValue) {
  return {
    type: 'SKINS_SORT_METHOD_CHANGE',
    payload: nextValue
  };
}
