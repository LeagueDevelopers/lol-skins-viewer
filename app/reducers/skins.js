import u from 'updeep';

const DEFAULT_STATE = {
  hasLoaded: false,
  summoner: undefined,
  ownedChampionIds: [],
  champions: [],
  skinMetadata: {},
  filters: {
    show: 'OWNED',
    name: ''
  },
  sortMethod: 'ALPHABETICAL'
};

export default function skins (state = DEFAULT_STATE, action) {
  switch (action.type) {
    case 'RECEIVE_SKINS':
      return u({ hasLoaded: true, ...action.payload }, state);
    case 'SKINS_FILTER_NAME_CHANGE':
      return u({ filters: { name: action.payload } }, state);
    case 'SKINS_FILTER_SHOW_CHANGE':
      return u({ filters: { show: action.payload } }, state);
    case 'SKINS_SORT_METHOD_CHANGE':
      return u({ sortMethod: action.payload }, state);
    default:
      return state;
  }
}
