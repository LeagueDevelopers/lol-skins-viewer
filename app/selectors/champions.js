import { createSelector } from 'reselect';
import sortBy from 'lodash.sortby';
import orderBy from 'lodash.orderby';
import Fuse from 'fuse.js';

const nameSelector = state => state.bindings.filters.name;

const showUnownedSkins = state => state.bindings.filters.show;

const sortMethod = state => state.bindings.sortMethod;

export const championsSelector = state => state.app.champions;

export const filterOptions = [
  { value: 'ALL', label: 'All Skins' },
  { value: 'OWNED', label: 'Owned Skins' },
  { value: 'UNOWNED', label: 'Unowned Skins' }
];

const filterChampionsByOwnership = createSelector(
  championsSelector,
  showUnownedSkins,
  (champions, show) => {
    switch (show) {
      case 'ALL':
        return champions;
      default:
      case 'OWNED':
        return champions.filter(s => s.ownership.owned);
      case 'UNOWNED':
        return champions.filter(s => !s.ownership.owned);
    }
  }
);

export const championsIndexSelector = createSelector(
  filterChampionsByOwnership,
  champions => new Fuse(champions, {
    keys: ['alias', 'name'],
    threshold: '0.1'
  })
);

export const sortingOptions = [
  { value: 'ALPHABETICAL', label: 'ALPHABETICAL' },
  { value: 'MASTERY', label: 'MASTERY' },
  { value: 'RP VALUE', label: 'RP VALUE' },
  { value: 'SKINS OWNED', label: 'SKINS OWNED' },
];

export const filteredChampions = createSelector(
  championsIndexSelector,
  nameSelector,
  (championIndex, name) => {
    if (!name || name === '') {
      return sortBy(championIndex.list, ['championName', 'name']);
    }
    return sortBy(championIndex.search(name), ['championName', 'name']);
  }
);

export const sortedChampions = createSelector(
  filteredChampions,
  sortMethod,
  (champions, sort) => {
    switch (sort) {
      default:
      case 'ALPHABETICAL':
        return champions;
      case 'MASTERY':
        return orderBy(champions, ['masteryLevel', 'masteryPoints'], ['desc', 'desc']);
      case 'RP VALUE':
        return orderBy(champions, ['rpValue'], ['desc']);
      case 'SKINS OWNED':
        return orderBy(champions, ['championOwnedCount'], ['desc']);
    }
  }
);
