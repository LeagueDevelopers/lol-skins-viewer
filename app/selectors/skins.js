import { createSelector } from 'reselect';
import flatMap from 'lodash.flatmap';
import sortBy from 'lodash.sortby';
import orderBy from 'lodash.orderby';
import Fuse from 'fuse.js';

const skinMetadataSelector = state => state.skins.skinMetadata;

const nameSelector = state => state.skins.filters.name;

const showUnownedSkins = state => state.skins.filters.show;

const sortMethod = state => state.skins.sortMethod;

const championsSelector = state => state.skins.champions;

export const skinsSelector = createSelector(
  championsSelector,
  skinMetadataSelector,
  (champions, skinMetadata) => flatMap(champions, c => {
    const championTotalCount = c.skins.length;
    const championOwnedCount = c.skins.filter(s => s.owned).length;
    const masteryLevel = c.masteryLevel;
    const masteryPoints = c.masteryPoints;
    return c.skins.map(s => {
      const metadata = skinMetadata[s.id];
      let rpValue = 9999;
      if (metadata && metadata.rpValue && metadata.rpValue > 0) {
        rpValue = metadata.rpValue;
      }
      return {
        ...s,
        championTotalCount,
        championOwnedCount,
        masteryLevel,
        masteryPoints,
        rpValue
      };
    });
  })
);

export const countOwnedSkins = createSelector(
  skinsSelector,
  skins => skins.filter(s => s.owned).length
);

export const filterOptions = [
  { value: 'ALL', label: 'All Skins' },
  { value: 'OWNED', label: 'Owned Skins' },
  { value: 'UNOWNED', label: 'Unowned Skins' }
];

const filterSkinsByOwnership = createSelector(
  skinsSelector,
  showUnownedSkins,
  (skins, show) => {
    switch (show) {
      case 'ALL':
        return skins;
      default:
      case 'OWNED':
        return skins.filter(s => s.owned);
      case 'UNOWNED':
        return skins.filter(s => !s.owned);
    }
  }
);

export const skinIndexSelector = createSelector(
  filterSkinsByOwnership,
  skins => new Fuse(skins, {
    keys: ['championName', 'name'],
    threshold: '0.1'
  })
);

export const sortingOptions = [
  { value: 'ALPHABETICAL', label: 'ALPHABETICAL' },
  { value: 'MASTERY', label: 'MASTERY' },
  { value: 'RP VALUE', label: 'RP VALUE' },
  { value: 'SKINS OWNED', label: 'SKINS OWNED' },
];

export const filteredSkins = createSelector(
  skinIndexSelector,
  nameSelector,
  (skinIndex, name) => {
    if (!name || name === '') {
      return skinIndex.list;
    }
    return skinIndex.search(name);
  }
);

export const sortedSkins = createSelector(
  filteredSkins,
  sortMethod,
  (skins, sort) => {
    switch (sort) {
      default:
      case 'ALPHABETICAL':
        return sortBy(skins, ['championName', 'name']);
      case 'MASTERY':
        return orderBy(skins, ['masteryLevel', 'masteryPoints', 'championName', 'name'], ['desc', 'desc', 'asc', 'asc']);
      case 'RP VALUE':
        return orderBy(skins, ['rpValue', 'championName', 'name'], ['desc', 'asc', 'asc']);
      case 'SKINS_OWNED':
        return orderBy(skins, ['championOwnedCount', 'championName', 'name'], ['desc', 'asc', 'asc']);
    }
  }
);
