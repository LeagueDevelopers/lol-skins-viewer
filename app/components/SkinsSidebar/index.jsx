import React, { PropTypes } from 'react';

import { filterOptions, sortingOptions } from 'selectors/skins';

import Dropdown from 'components/Dropdown';
import Input from 'components/Input';
import RadioInput from 'components/RadioInput';

import style from './index.scss';

const SkinsSidebar = ({ count, sortMethod, filters, changeNameFilter, changeShowFilter, changeSortMethod }) => (
  <div className={style.sidebar}>
    <div className={style.summary}>
      <span className={style.skinsCount}>{count}</span>
      <span className={style.skinsText}>Total Skins</span>
    </div>
    <div className={style.filters}>
      <Input type="search" value={filters.name} onChange={changeNameFilter} placeholder="Search" />
      <Dropdown options={filterOptions} value={filters.show} onChange={changeShowFilter} />
      <RadioInput options={sortingOptions} value={sortMethod} onChange={changeSortMethod} />
    </div>
  </div>
  );

SkinsSidebar.propTypes = {
  count: PropTypes.number.isRequired,
  filters: PropTypes.object.isRequired,
  sortMethod: PropTypes.string.isRequired,
  changeNameFilter: PropTypes.func.isRequired,
  changeShowFilter: PropTypes.func.isRequired,
  changeSortMethod: PropTypes.func.isRequired
};

export default SkinsSidebar;
