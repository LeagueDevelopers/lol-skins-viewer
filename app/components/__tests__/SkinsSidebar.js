import React from 'react';
import { shallow, mount } from 'enzyme';

import { sortingOptions } from 'selectors/skins';

import SkinsSidebar from '../SkinsSidebar';

jest.useFakeTimers();

jest.mock('lodash.debounce', () => {
  let debounced;
  const debounce = (fn, timeout) => {
    const res = val => {
      debounced = () => fn(val);
      setTimeout(debounced, timeout);
    };
    res.flush = () => debounced();
    return res;
  };
  return debounce;
});

describe('<SkinsSidebar />', () => {
  it('should render correctly', () => {
    const props = {
      rpFiltered: 1234,
      count: 50,
      filters: {
        show: 'OWNED'
      },
      sortMethod: sortingOptions[0].value,
      changeShowFilter: jest.fn(),
      changeSortMethod: jest.fn()
    };

    const wrapper = shallow(<SkinsSidebar {...props} />);

    expect(wrapper.get(0)).toMatchSnapshot();
  });

  it('should toggle between RP Total and Skins Count', () => {
    ['ALL', 'OWNED', 'UNOWNED'].forEach((type) => {
      const props = {
        rpFiltered: 1234,
        count: 50,
        filters: {
          show: type
        }
      };
      const wrapper = shallow(<SkinsSidebar {...props} />);

      expect(wrapper.first().childAt(0).props().children).toBe(`${type} Skins`);
      expect(wrapper.first().childAt(0).props().value).toBe(props.count);

      wrapper.instance().toggleSummary();

      expect(wrapper.first().childAt(0).props().children).toBe('Estimated RP Value');
      expect(wrapper.first().childAt(0).props().value).toBe(props.rpFiltered);
    });
  });

  it('should debounce search change calls', () => {
    const handleNameChange = jest.fn();

    const wrapper = mount(
      <SkinsSidebar filters={{ name: 'Jinx' }} changeNameFilter={handleNameChange} />
    );
    const searchInput = wrapper.childAt(1).childAt(0);

    handleNameChange.mockImplementation(name => wrapper.setProps({ filters: { name } }));

    expect(searchInput.props().value).toBe('Jinx');

    wrapper.instance().handleSearchChange('Leona');

    expect(searchInput.props().value).toBe('Leona');
    expect(handleNameChange).not.toHaveBeenCalled();

    jest.runOnlyPendingTimers();

    expect(handleNameChange).toHaveBeenCalledWith('Leona');
  });

  it('should immediately flush pending debounced calls on unmount', () => {
    const handleNameChange = jest.fn();

    const wrapper = mount(
      <SkinsSidebar filters={{ name: 'Jinx' }} changeNameFilter={handleNameChange} />
    );

    const searchInput = wrapper.childAt(1).childAt(0);

    expect(searchInput.props().value).toBe('Jinx');

    wrapper.instance().handleSearchChange('Leona');

    expect(searchInput.props().value).toBe('Leona');
    expect(handleNameChange).not.toHaveBeenCalled();

    wrapper.unmount();

    expect(handleNameChange).toHaveBeenCalledWith('Leona');
  });
});
