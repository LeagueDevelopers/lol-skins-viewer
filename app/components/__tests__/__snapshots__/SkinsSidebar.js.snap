exports[`<SkinsSidebar /> should render correctly 1`] = `
<div
  className="sidebar">
  <SummaryRing
    className="ring"
    onClick={[Function]}
    value={50}>
    OWNED Skins
  </SummaryRing>
  <div
    className="filters">
    <Input
      onChange={[Function]}
      placeholder="Search"
      type="search"
      value="" />
    <Dropdown
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "All Skins",
            "value": "ALL",
          },
          Object {
            "label": "Owned Skins",
            "value": "OWNED",
          },
          Object {
            "label": "Unowned Skins",
            "value": "UNOWNED",
          },
        ]
      }
      value="OWNED" />
    <RadioInput
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "ALPHABETICAL",
            "value": "ALPHABETICAL",
          },
          Object {
            "label": "MASTERY",
            "value": "MASTERY",
          },
          Object {
            "label": "RP VALUE",
            "value": "RP VALUE",
          },
          Object {
            "label": "SKINS OWNED",
            "value": "SKINS OWNED",
          },
        ]
      }
      value="ALPHABETICAL" />
  </div>
</div>
`;
